src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java

```diff
@@ -0,0 +1,446 @@
+        // which prevents them from lagging the detection
+        this.frameWindow.reset().fill(0);
+        this.encodeWindow.reset().fill(0);
+
+        // reset the encoder states
+        while (this.encodeModel.states().hasRemaining())
+            this.encodeModel.states().putFloat(0);
+    }
+
+    /**
+     * processes a frame of audio.
+     * @param context the current speech context
+     * @param buffer  the audio frame to detect
+     * @throws Exception on error
+     */
+    public void process(SpeechContext context, ByteBuffer buffer)
+            throws Exception {
+        // run the current frame through the detector pipeline
+        sample(context, buffer);
+
+        // on deactivation, see if a keyword was detected
+        if (!context.isActive() && this.isActive)
+            detect(context);
+
+        this.isActive = context.isActive();
+    }
+
+    private void sample(SpeechContext context, ByteBuffer buffer) {
+        // process all samples in the frame
+        buffer.rewind();
+        while (buffer.hasRemaining()) {
+            // convert to float and clip the 16-bit sample
+            float sample = (float) buffer.getShort() / Short.MAX_VALUE;
+            sample = Math.max(-1f, Math.min(sample, 1f));
+
+            // run a pre-emphasis filter to balance high frequencies
+            // and eliminate any dc energy
+            float nextSample = sample;
+            sample -= this.preEmphasis * this.prevSample;
+            this.prevSample = nextSample;
+
+            // process the sample
+            // . write it to the sample sliding window
+            // . run the remainder of the detection pipeline if active
+            // . advance the sliding window by the hop length
+            this.sampleWindow.write(sample);
+            if (this.sampleWindow.isFull()) {
+                if (context.isActive())
+                    analyze(context);
+                this.sampleWindow.rewind().seek(this.hopLength);
+            }
+        }
+    }
+
+    private void analyze(SpeechContext context) {
+        // apply the windowing function to the current sample window
+        for (int i = 0; i < this.fftFrame.length; i++)
+            this.fftFrame[i] = this.sampleWindow.read() * this.fftWindow[i];
+
+        // compute the stft
+        this.fft.realForward(this.fftFrame);
+
+        filter(context);
+    }
+
+    private void filter(SpeechContext context) {
+        // decode the FFT outputs into the filter model's inputs
+        // . compute the magnitude (abs) of each complex stft component
+        // . the first and last stft components contain only real parts
+        //   and are stored in the first two positions of the stft output
+        // . the remaining components contain real/imaginary parts
+        this.filterModel.inputs(0).rewind();
+        this.filterModel.inputs(0).putFloat(this.fftFrame[0]);
+        for (int i = 1; i < this.fftFrame.length / 2; i++) {
```

Comment:
I know this loop is identical in the wakeword code, but isn't it more conventional to loop until `this.fftFrame.length` and increment by 2 on each iteration than to do all the inline multiplication?

Reply:
I don't think there is a java convention here, we'll leave it consistent with the wakeword detector for now